Get gcda and gcno files from containers.

ArtiomDivak opened the pull request:
When running tmt gcda and gcno files are created. This files are necessary for later code coverage report creation. So before the container destroyed we need to extract this files to the local machine.

Related: https://github.com/containers/bluechi/issues/397

ArtiomDivak committed:
Get gcda and gcno files from containers.

[WIP]
When running tmt gcda and gcno files are created. This files necessary for later code coverage.
So before the container destroyed we need to extract this files to the local machine.

Related: https://github.com/containers/bluechi/issues/397
Signed-off-by: Artiom Divak <adivak@redhat.com>

mwperina commented on tests/bluechi_test/container.py:
```diff
@@ -63,6 +63,13 @@ def gather_journal_logs(self, data_dir: str) -> None:
+        gcno_file_location = "/usr/share/bluechi-coverage/"
+        gcda_file_location = "/root/rpmbuild/BUILD/bluechi-0.6.0/redhat-linux-build/src"
+
+        self.get_file(gcno_file_location, data_dir)
```
Could you please create `coverage` subdirectory under `data_dir` and put coverage files there?

mwperina commented on tests/bluechi_test/test.py:
```diff
@@ -166,6 +175,7 @@ def run(self, exec: Callable[[BluechiControllerContainer, Dict[str, BluechiNodeC
             exec(ctrl_container, node_container)
         finally:
             try:
+                self.gather_coverage(ctrl_container, node_container)
```
I think we should gather coverage only when we build RPMs with coverage support. So I'd probably add following:

1. Define tmt environment variable `WITH_COVERAGE` with default value `no` similarly as in https://github.com/containers/bluechi/pull/488/files#diff-ec5e35f9a00ac177713459ba4f15ab7a84faa6d69adb1d05a2134f2f15ea0dc5R10
2. Create `with_coverage` fixture which will expose `WITH_COVERAGE` value to the test code similary as in https://github.com/containers/bluechi/blob/main/tests/bluechi_test/fixtures.py#L50
3. Run coverage gathering only when `with_coverage` returns `True`

mwperina commented on tests/containers/integration-test-local:
```diff
@@ -14,6 +14,7 @@ RUN dnf install -y --repo bluechi-rpms \
         bluechi-ctl \
         bluechi-ctl-debuginfo \
         bluechi-selinux \
+        bluechi-coverage \
```
Similar to above, I'd pass `WITH_COVERAGE` value as an argument to container build similar as in https://github.com/containers/bluechi/pull/488/files#diff-fe5746460187f3d335a236fa749d7516f1e342a307150d3e7c3f4aec7183203cR3 and  https://github.com/containers/bluechi/pull/488/files#diff-66c84c667aa7801b2592d803c12fd54c2131c7a17dc74518221f9be618f8a571R20
Then we can install `bluechi-coverage` package only when needed

mwperina commented on tests/bluechi_test/container.py:
```diff
@@ -63,6 +63,15 @@ def gather_journal_logs(self, data_dir: str) -> None:
 
+    def gather_coverage(self, data_dir: str) -> None:
+        gcno_file_location = "/usr/share/bluechi-coverage/"
+        gcda_file_location = "/root/rpmbuild/BUILD/bluechi-0.6.0/redhat-linux-build/src"
```
Wouldn't gcda files be created in `/usr/libexec` for our daemons and in `/usr/bin` for `bluechictl`?

mwperina commented on tests/bluechi_test/container.py:
```diff
@@ -63,6 +63,15 @@ def gather_journal_logs(self, data_dir: str) -> None:
+        gcno_file_location = "/usr/share/bluechi-coverage/"
+        gcda_file_location = "/root/rpmbuild/BUILD/bluechi-0.6.0/redhat-linux-build/src"
+
+        result, output = self.exec_run(f"ls -la {gcda_file_location}", tty=True)
```
What is the `ls -al` output for? Don't you know the exact location and name of each gcda file?

ArtiomDivak replied:
Form what I have found the gcda location is in "/root/rpmbuild/BUILD/bluechi-0.6.0/redhat-linux-build/src" and there is no other path to the gcda. the `ls -la` command is to check if the location exist because in some nodes there is no such location(and no gcda files). Maybe I need a better command to check if dir exist.

mwperina replied:
Hmm, so if  `.gcda` files are created by default in the same location as object files https://splichal.eu/gccsphinx-final/html/gcc/gcov/brief-description-of-gcov-data-files.html then we need additional PR, which would contain:

1. Modify `meson.build` to add `-fprofile-dir=/var/tmp/bluechi-coverage` to compiler when building with coverage created
2. Modify `bluechi.spec.in` which would created `/var/tmp/bluechi-coverage` as a part of `bluechi-coverage` RPM installation

With those changes above you could just gather `.gcda` files from `/var/tmp/bluechi-coverage` without any detection magic

mwperina commented on tests/bluechi_test/test.py:
```diff
@@ -116,6 +116,16 @@ def gather_logs(self, ctrl: BluechiControllerContainer, nodes: Dict[str, Bluechi
+            ctrl.gather_coverage(self.tmt_test_data_dir)
+
+        for _, node in nodes.items():
+            logger.debug("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
```
Maybe worth changing to:

```python
logger.debug(f"Gathering coverage files from node '{node.node_name}'")
```

ArtiomDivak replied:
This  was added for my debugging I will soon remove it.

ArtiomDivak committed:
Add path for gcda files

Added arguments for meson so the gcda files will be saved in /var/tmp/bluechi-coverage.
Also bluechi.spec.in will add the /var/tmp/bluechi-coverage.

Signed-off-by: Artiom Divak <adivak@redhat.com>

Add path for gcda files

Added arguments for meson so the gcda files will be saved in /var/tmp/bluechi-coverage.
Also bluechi.spec.in will add the /var/tmp/bluechi-coverage.

Signed-off-by: Artiom Divak <adivak@redhat.com>

mwperina commented on bluechi.spec.in:
```diff
@@ -214,6 +214,7 @@ will be used during integration tests when creating code coverage report.
 %files coverage
 %license LICENSE
 %{_datadir}/bluechi-coverage/*.gcno
+%/var/tmp/bluechi-coverage/
```
Directories should be using `%dir` macro and you should be using `%{_localstatedir}` instead of `/var` path:

```
%dir %{_localstatedir}/tmp/bluechi-coverage/
```

mwperina commented on meson.build:
```diff
@@ -31,6 +31,7 @@ endif
 with_coverage = get_option('with_coverage')
 if with_coverage == 'true'
    add_project_arguments('-coverage', language : 'c')
+   add_project_arguments('-fprofile-dir=/var/tmp/bluechi-coverage', language : 'c')
```
You should not use hardcoded path, but rather using specific options for directories:

```
 add_project_arguments(
    '-fprofile-dir=' + join_paths(get_option('prefix'), get_option('localstatedir'), 'tmp', 'bluechi-coverage'), 
    language : 'c')
```

mwperina commented on tests/README.md:
````diff
@@ -86,6 +86,16 @@ WITH_VALGRIND=1 tmt run -v
+```
+
+the  coverage will get all the gcda and gcno files to the `bluechi-coverage` dir under `data` dir.
+
````
Hmm, I'd probably mention coverage options in README only after the last step, where we will have the coverage report generated

mwperina commented on tests/containers/integration-test-snapshot:
```diff
@@ -2,6 +2,10 @@ FROM quay.io/bluechi/integration-test-base:latest
 
+ARG with_coverage
+
+ARG with_coverage
```
Please remove duplicate argument definition

mwperina commented on tests/containers/integration-test-snapshot:
```diff
@@ -19,4 +23,13 @@ RUN dnf install \
+        bluechi-coverage && \
+    dnf -y clean all; \
+    fi
+
```
This is wrong, in `integration-test-snapshot` container we need to install everything from bluechi-snapshot repo, we are not using any local repo in this container

mwperina commented on tests/containers/integration-test-local:
```diff
@@ -19,4 +21,13 @@ RUN dnf install --repo bluechi-rpms \
+        --nogpgcheck \
+        --nodocs \
+        bluechi-coverage && \
+    dnf -y clean all; \
```
Ah, sorry, I missed that before :-(
We are cleaning DNF cache on line 20, so it would be better to move the IF section before that line and remove clean from IF section, so we clean only once

mwperina commented on tests/containers/integration-test-snapshot:
```diff
@@ -19,4 +21,12 @@ RUN dnf install \
+        --nogpgcheck \
+        --nodocs \
+        bluechi-coverage && \
+    dnf -y clean all; \
```
Same here, please clean only once

mwperina commented on tests/containers/integration-test-local:
```diff
@@ -1,8 +1,20 @@
+        --nodocs \
+        bluechi-coverage; \
+    fi
+
```
Please install bluechi-coverage only after bluechi installation and before running clean all

mwperina commented on tests/containers/integration-test-snapshot:
```diff
@@ -2,8 +2,17 @@ FROM quay.io/bluechi/integration-test-base:latest
+        --nodocs \
+        bluechi-coverage;  \
+    fi
+
```
Same as above

mwperina reviewed (approved):
LGTM

engelmi commented on meson.build:
```diff
@@ -31,6 +31,9 @@ endif
 if with_coverage == 'true'
    add_project_arguments('-coverage', language : 'c')
+    add_project_arguments(
+    '-fprofile-dir=' + join_paths(get_option('prefix'), get_option('localstatedir'), 'tmp', 'bluechi-coverage'), 
```
This has probably been discussed before, but I wonder why we need to add the prefix etc. to the path and not simply use a static one like `/tmp/bluechi-coverage`? The coverage is for test purposes only, so I think its better to be static here - so we can rely on it later, e.g. when collecting those files. Using options this seems harder.

mwperina replied:
Personally I'd like to be consistent and make that working even when building for non-default prefix

engelmi replied:
Ok, I am fine with that. Just wanted to ask :)

engelmi commented on tests/bluechi_test/test.py:
```diff
@@ -120,6 +122,19 @@ def gather_logs(self, ctrl: BluechiControllerContainer, nodes: Dict[str, Bluechi
+            ctrl.gather_coverage(data_coverage_dir)
+
+        for _, node in nodes.items():
+            ctrl.gather_coverage(data_coverage_dir)
```
```suggestion
            node.gather_coverage(data_coverage_dir)
```

mwperina replied:
Good catch, I totally missed it :-(

mwperina reviewed (approved):
LGTM

engelmi reviewed (approved):
LGTM
